nomad/event_endpoint.go

```diff
@@ -0,0 +1,210 @@
+package nomad
+
+import (
+	"context"
+	"io"
+	"time"
+
+	"github.com/hashicorp/go-msgpack/codec"
+	"github.com/hashicorp/nomad/helper"
+	"github.com/hashicorp/nomad/helper/uuid"
+	"github.com/hashicorp/nomad/nomad/stream"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+type Event struct {
+	srv *Server
+}
+
+func (e *Event) register() {
+	e.srv.streamingRpcs.Register("Event.Stream", e.stream)
+}
+
+func (e *Event) stream(conn io.ReadWriteCloser) {
+	defer conn.Close()
+
+	var args structs.EventStreamRequest
+	decoder := codec.NewDecoder(conn, structs.MsgpackHandle)
+	encoder := codec.NewEncoder(conn, structs.MsgpackHandle)
+
+	if err := decoder.Decode(&args); err != nil {
+		handleJsonResultError(err, helper.Int64ToPtr(500), encoder)
+		return
+	}
+
+	// forward to appropriate region
+	if args.Region != e.srv.config.Region {
+		err := e.forwardStreamingRPC(args.Region, "Event.Stream", args, conn)
+		if err != nil {
+			handleJsonResultError(err, helper.Int64ToPtr(500), encoder)
+		}
+		return
+	}
+
+	// ACL check
+	// TODO(drew) ACL checks need to be per topic
+	// All Events        Management
+	// System Events     Management
+	// Node Events       NamespaceCapabilityReadEvents
+	// Job/Alloc Events  NamespaceCapabilityReadEvents
+	if aclObj, err := e.srv.ResolveToken(args.AuthToken); err != nil {
+		handleJsonResultError(err, nil, encoder)
+		return
+	} else if aclObj != nil && !aclObj.IsManagement() {
+		handleJsonResultError(structs.ErrPermissionDenied, helper.Int64ToPtr(403), encoder)
+		return
+	}
+
+	// TODO(drew) handle streams without ACLS
+	reqToken := args.AuthToken
+	if reqToken == "" {
+		// generate a random request token
+		reqToken = uuid.Generate()
+	}
+	subReq := &stream.SubscribeRequest{
+		Token:  reqToken,
+		Topics: args.Topics,
+		Index:  uint64(args.Index),
+	}
+	publisher, err := e.srv.State().EventPublisher()
+	if err != nil {
+		handleJsonResultError(err, helper.Int64ToPtr(500), encoder)
+		return
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// start subscription to publisher
+	subscription, err := publisher.Subscribe(subReq)
+	if err != nil {
+		handleJsonResultError(err, helper.Int64ToPtr(500), encoder)
+		return
+	}
+	defer subscription.Unsubscribe()
+
+	ndJsonCh := make(chan *stream.NDJson)
+	errCh := make(chan error)
+
+	jsonStream := stream.NewNDJsonStream(ndJsonCh, 30*time.Second)
+	jsonStream.Run(ctx)
+
+	// goroutine to detect remote side closing
+	go func() {
+		if _, err := conn.Read(nil); err != nil {
+			// One end of the pipe explicitly closed, exit
+			cancel()
+			return
+		}
+		select {
+		case <-errCh:
```

Comment:
Why are we receiving on the errCh here?

I know this is the pattern used in Monitor, but is checking the error returned by `conn.Read` above even necessary? Seems like in streaming APIs where we never expect to read anything, `conn.Read` returning anything at all is an indication we should shut things down?

Reply:
I don't think we need receive on errCh here, removing. I think we want to try to read from the conn just to make sure the conneciton hasn't closed?